### PR TITLE
Support Pull Request and issue template

### PR DESCRIPTION
**Pull request by ganmacs**
Ref #1118 

Support Pull  request and issues template.

### CODE REVIEW

File: github/template.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/github/hub/utils"
+)
+
+const (
+	pullRequestTempalte = "PULL_REQUEST_TEMPLATE"
+	issueTempalte       = "ISSUE_TEMPLATE"
+	githubTemplateDir   = ".github"
+)
+
+func GetPullRequestTempalte() string {
```
Comment (scompt) [inline github/template.go]: There's a typo here. Should be `GetPullRequestTemplate `

File: github/template.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	return getGithubTemplate(pullRequestTempalte)
+}
+
+func GetIssueTempalte() string {
```
Comment (scompt) [inline github/template.go]: There's a typo here. Should be `GetIssueTemplate `

**Comment by dawehner** [comment]
It would be quite nice to have that

**Comment by jeroenvisser101** [comment]
@ganmacs is there anything keeping this from being merged?

**Comment by NickLaMuro** [comment]
@jeroenvisser101 So I have used the custom homebrew recipe below to install this from @ganmacs branch:

```ruby
class Hub < Formula
  desc "Add GitHub support to git on the command-line"
  homepage "https://hub.github.com/"
  url 'https://github.com/ganmacs/hub.git', :using => :git, :branch => 'support-github-template'
  version '2'

  option "without-completions", "Disable bash/zsh completions"

  depends_on "go" => :build

  def install
    system "script/build", "-o", "hub"
    bin.install "hub"
    man1.install Dir["man/*"]

    if build.with? "completions"
      bash_completion.install "etc/hub.bash_completion.sh"
      zsh_completion.install "etc/hub.zsh_completion" => "_hub"
    end
  end

  test do
    HOMEBREW_REPOSITORY.cd do
      assert_equal "bin/brew", shell_output("#{bin}/hub ls-files -- bin").strip
    end
  end
end
```

Pretty much the same as the [original formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/hub.rb), so if you use homebrew and are on a mac, you can install it by saving the file above and just running `brew install name_of_this_file.rb`.

That said, there are a few issues that I have noticed since using this that would probably make it not mergable:

* The PR template is meant to be just the description section of the PR, but when it isn't a single commit PR, the template is added to the editor as is, which makes the first line of the template the title instead of the start of the body.
* With a single commit PR, the following happens:
  - If it is just a single line commit, the commit messages is appended to the first line of the template instead of making it the first line and adding two new lines in between
  - If it is a multi-line commit message, same problem, but it spans multiple lines until the last line of the commit message, where the last line of the commit message and the first line of the template are the same.
* Possible typo [here](https://github.com/github/hub/pull/1145/files#diff-319d83eedfd5d5f46fead682343297f7R281) (pedantic, so I would say it doesn't need to be address)
* No tests added for this (arguably optional, but based on the current verbosity of the test suite, I doubt it)

**Comment by jeroenvisser101** [comment]
@NickLaMuro thanks – the problems you mention I can see as well. I'll try using the homebrew recipe you sent, thanks!

**Comment by NickLaMuro** [comment]
To anyone hoping to use this before it is merged, I have made some updates to @ganmacs's PR in the following:

https://github.com/ganmacs/hub/pull/1

I have no expectations that this will or will not get merged in, but I am providing it as a stop gap for those that would like to use this feature (as is) before it is merged and released (or redone entirely for all I know).

For those on OSX, the above homebrew recipe will work, but requires a few updates:

* Make sure the file is named `hub.rb`.  I failed to mention that before (because I forgot), but Homebrew will barf if you don't have that same name as the class name of the file.
* Update the url in the formula from "https://github.com/ganmacs/hub.git" to "https://github.com/NickLaMuro/hub.git" to get my updates.

**Comment by jeroenvisser101** [comment]
@NickLaMuro that's great! You might want to consider opening up a PR from `NickLaMuro:support-github-template` to `github:master` yourself, as it'd probably get merged sooner. Thanks for your effort, it's greatly appreciated!

**Comment by NickLaMuro** [comment]
@ganmacs These changes are excellent.  I am humbled, and my lame attempt at `go` has been out-done by what you have done here!

Thanks for this.  I learned at lot!

**Comment by jeroenvisser101** [comment]
@NickLaMuro I really appreciate your attempt, and since you started all this, I thank you a lot! I don't think it was a lame attempt at all :)

**Comment by jeroenvisser101** [comment]
@mislav is this something that we can get into master?

**Comment by mislav** [comment]
This looks great. Thanks @ganmacs for doing the hard work, @NickLaMuro for in-depth review and contribution, and @jeroenvisser101 for 👀 and pings. I'll take this out for a spin and see how it feels in the text editor, and then merge if I don't find any problems. 👍

**Comment by jez** [comment]
Hey @mislav, have you had some time to poke around with it? What are your thoughts?

I've been poking around myself, and here are some things that I noticed:

- [ ] It doesn't handle case insensitivity ([relevant docs][1])
- [ ] It doesn't handle `.txt` extensions ([relevant docs][1])

It's a simple change to fix these; in fact I've already implemented it here: [jez/hub@2b777f1][2].

As far as I can tell though, it works Well Enough™, so we could probably merge the current version and address these and those that @NickLaMuro mentioned in PR's to follow.

[1]: https://help.github.com/articles/creating-an-issue-template-for-your-repository/
[2]: https://github.com/jez/hub/commit/2b777f1bd6718b31c0c18706b82973557acfb788

**Comment by jeroenvisser101** [comment]
@jez great points, and I agree, if nothing serious comes up, merge and follow up with improvements/fixes :)

**Comment by jez** [comment]
I found and fixed another issue: there was an error with the testing fixture code which was causing a test that should have failed to not fail. I've fixed the failing test, and the faulty fixture code it relied on in [jez/hub@499bfa8][1].

[1]: https://github.com/jez/hub/commit/499bfa821a88931ea9c96bce3c7c8b8005ecf6a4

**Comment by mislav** [comment]
@jez: Thanks for testing and fixes! Feel free to create a branch that adds your fixes on top of `ganmacs:support-github-template`. I can that pull in both this PR and your fixes in the same go.

**Comment by jez** [comment]
I already have a branch on my fork: [`case-insensitive-matching`][1]. Is that what you wanted, or do you want a separate PR?

[1]: https://github.com/jez/hub/tree/case-insensitive-matching

**Comment by jez** [comment]
Bump :D

**Comment by rpechayr** [comment]
@jez I guess you work (on top of  anmacs:support-github-template) need to be in a PR of this repo so that everything can be reviewed and merged altogether.  

This work is actually quite important for my team because we want to enforce a Definition of Done for every PR thought a PR template and we want to be able to do that using hub :)

**Comment by jez** [comment]
@rpechayr, @mislav: I've created a separate PR (see reference above). It includes all the changes here as well as my changes. It's ready to be merged. Any further conversation can probably happen in that PR.

**Comment by extemporalgenome** [comment]
I've been using this patch for a while now, and it does have one issue: if PULL_REQUEST_TEMPLATE.md contains markdown headers (starting with `#`), those lines get stripped out when creating the PR. Similar behavior will occur for any markup language that uses lines starting with the git comment char. To properly handle that, the pull-request command would likely need to suppress insertion of typical git instruction lines, while also suppressing comment-stripping.

**Comment by jez** [comment]
__TL;DR__: That's a feature not a bug. Here's what I mean, and what you can do to use the feature as intended.

@extemporalgenome This is actually expected behavior, and hub has a built-in way of handling it using git's `core.commentChar`. From the manual:

```
core.commentChar
    Commands such as commit and tag that lets you edit messages
    consider a line that begins with this character commented,
    and removes them after the editor returns (default #).
```

If you set it to something like `%`, this will free up `#` to be used as a Markdown ATX header. You can set this option globally (`git config --global`), local to a specific repo (`git config`), or per-command (`git -c core.commentChar='%' ...`). Because most Git syntax highlighters only consider `#` to be a comment character (and I don't want to lose that visual feedback when making commits), I've opted for the 3rd invocation, using these aliases:

```shell
# We want to use '#' as a markdown character, so let's use '%' for comments
alias hubmdpr="hub -c core.commentChar='%' pull-request"
alias hubmdis="hub -c core.commentChar='%' issue create"
```

Additionally, I wrote a Vim plugin that plays nicely with custom comment characters in issue and pull request buffers: [jez/vim-github-hub][1].

[1]: https://github.com/jez/vim-github-hub

File: github/template.go (hunk 2 above)
Comment (dharmik0070) [reply github/template.go]: Feature

File: github/template.go (hunk 1 above)
Comment (dharmik0070) [reply github/template.go]: Feature

Comment (dharmik0070) [reply github/template.go]: hhhh
Feature

**Pull request closed without merging**